### PR TITLE
Added support for xeus-cling.

### LINKED ISSUE #150: Add C++ support to libretexts pages
We now have C++ working in default-env. We need to setup Libretexts pages to use default-env and enable the C++ kernel support in the pages.

### PR DESCRIPTION
**Pull request by moorepants**
@LibreTexts/jupyter-team Can I get a review here? I'm also not sure how I'd try it out myself.

Fixes https://github.com/LibreTexts/metalc/issues/150

### CODE REVIEW

**Comment by moorepants** [comment]
How does the syntax highlighting work? Are one of these config variables associated with the syntax highlighting language for the text box?

**Comment by moorepants** [comment]
Here is some info from the xeus example notebook:
```
 "metadata": {
  "kernelspec": {
   "display_name": "C++14",
   "language": "C++14",
   "name": "xcpp14"
  },
  "language_info": {
   "codemirror_mode": "text/x-c++src",
   "file_extension": ".cpp",
   "mimetype": "text/x-c++src",
   "name": "c++",
   "version": "14"
  }
```

**Comment by moorepants** [comment]
This is working for me locally.

**Comment by tannavee** [comment]
> How does the syntax highlighting work? Are one of these config variables associated with the syntax highlighting language for the text box?

I believe the syntax highlighting is adjusted from another file. I can work on adding that commit right now.

**Comment by moorepants** [comment]
Please suggest changes instead of pushing to my branch.

File: src/scripts/activateThebelab.js [hunk 1]
```diff
@@ -34,6 +34,11 @@ const getConfig = (language) => {
   };
 
   switch (language) {
+    case 'cpp':
+      config.binderOptions.repo = 'jupyter-xeus/xeus-cling';
+      config.binderOptions.ref = 'stable';
```
Comment (moorepants) [inline src/scripts/activateThebelab.js:39]: I chose the stable branch because that is what try.jupyter.org is using.

Comment (tannavee) [reply src/scripts/activateThebelab.js:39]: makes sense, I meant to push to a new branch to test, and have reverted it to the previous version

**Comment by moorepants** [comment]
I wonder if it is sufficient to support the latest version of C++. I'm not familiar with forwards compatible C++ code is, but my guess is that is likely is. So we could just use C++17 (is that the latest), if that works. Then we'd only need one C++ option.

File: src/scripts/dialogConfig.js [hunk 2]
```diff
@@ -11,6 +11,7 @@ const languageDictionary = {
   R: 'r',
   Octave: 'octave',
   SageMath: 'sagemath',
+  'C++': 'cpp',
```
Comment (tannavee) [inline src/scripts/dialogConfig.js]:
```suggestion
  'C++': 'text/x-c++src',
```

File: src/scripts/dialogConfig.js [hunk 3]
```diff
@@ -19,6 +20,7 @@ const dataLanguageDictionary = {
   r: 'R',
   octave: 'Octave',
   sagemath: 'SageMath',
+  cpp: 'C++',
```
Comment (tannavee) [inline src/scripts/dialogConfig.js]:
```suggestion
  'text/x-c++src': 'C++',
```

File: src/scripts/dialogConfig.js [hunk 4]
```diff
@@ -59,6 +61,10 @@ const editScriptAreaHTML = (language = 'python', code = null, output = null) =>
       code: 'print(\'Hello world!\')',
       output: 'Hello world!',
     },
+    cpp: {
```
Comment (tannavee) [inline src/scripts/dialogConfig.js]:
```suggestion
    'text/x-c++src': {
```

File: src/scripts/activateThebelab.js [hunk 5]
```diff
@@ -34,6 +34,11 @@ const getConfig = (language) => {
   };
 
   switch (language) {
+    case 'cpp':
```
Comment (tannavee) [inline src/scripts/activateThebelab.js]:
```suggestion
    case 'text/x-c++src':
```

**Comment by moorepants** [comment]
This works as is, so I think we should merge unless anyone has any comments. We can add other C++ versions and get syntax highlighting working later.

**Review by tannavee — COMMENTED** [review]
I think C++14 should be good for the time being. I tested everything out with the suggestions I left, and syntax highlighting works with it, so I believe we should be good on that front.

File: src/scripts/activateThebelab.js (hunk 5 above)
Comment (moorepants) [reply src/scripts/activateThebelab.js]: So should all of the languages have this type of change?

Comment (tannavee) [reply src/scripts/activateThebelab.js]: Yes, Celine and I were discussing perhaps creating a new js file to assign key values for the language and codemirror values so its clean. Or adding a section to an existing file. We didn't want to make changes to anything else just to make sure that everything is working.

**Comment by moorepants** [comment]
Thanks @tannavee I've changed to your suggestion. Testing locally works.

File: src/scripts/activateThebelab.js (hunk 5 above)
Comment (moorepants) [reply src/scripts/activateThebelab.js]: Sounds good.

**Review by tannavee — APPROVED** [review]

**Comment by celine168** [comment]
I wanted to preemptively try testing the insertion of code on LibreTexts because I wanted to make sure it wasn't just a problem on the development environment.
I opened an issue here: https://github.com/LibreTexts/ckeditor-binder-plugin/issues/83
The editor right now sends whatever is within the HTML to the kernel for it to execute, and it causes issues in parsing `< >` (brackets).

**Comment by moorepants** [comment]
So should we wait? I though the code mirror thing would handle those html parsing issues.

**Comment by celine168** [comment]
Tannavee and I figured out a solution that seems to work. I could suggest it. 
The issue is not on CodeMirror's side, it deals with how we are inserting code (which contains symbols) into an HTML editor which needs to escape these symbols and not treat them as HTML. There are more HTML entities that exist that we may want to consider though.

**Comment by celine168** [comment]
I made a commit to a new branch for reference: https://github.com/LibreTexts/ckeditor-binder-plugin/tree/fix/cpp-entities

**Comment by pbmacsodak** [comment]
Mornin Jason...been following the progress, but not sure what the current status is. Can you give me a bit more detail about where the implementation is at?? Just trying to figure out if I need to be thinkin of plan B, or if by the time I get to this in a bout a week we will be good to go?? Pat

**Comment by moorepants** [comment]
We've got C++ working in the editor, I'm not sure if it is available yet on the main libretexts site. @celine168 Does this automatically happen? Or do we need to ask @Miniland1333 to do something?

@pbmacsodak One issue we are facing is this HTML character escape issue, which can be annoying. Note that you'll need to get familiar with the intricacies of xeus-cling as there will be some unusual things since it is a C++ interpreter.

**Comment by Miniland1333** [comment]
For testing, files are manually uploaded test.libretexts.org using SFTP or some other similar transfer software. This will allow for testing on query.libretexts.org since it uses this staging/testing code. Someone on the Jupyterteam should be able to give you a copy of the credentials, or they can upload the built file for you to test.libretexts.org.

For production, master is automatically deployed using a Github action to all of the libraries.

**Comment by pbmacsodak** [comment]
I apologize, but I am confused. I am a libretexts author, desiring to use
Jupyter to include C++ code into my libretexts files, and be able to
execute it the way that other programming languages are implemented. I am
not sure which files need to be loaded to test.libretexts.org....or what is
it I need to do on query.libretexts.org. Is there something I need to do on
my end...I am not sure what the response means.
Pat

**Comment by celine168** [comment]
The current changes to the C++ kernel are on staging/query.libretexts.org. I'll create a PR from staging to master.

**Comment by celine168** [comment]
@pbmacsodak, you should be able to insert C++ code into LibreTexts after [this PR](https://github.com/LibreTexts/ckeditor-binder-plugin/pull/87) gets merged! query.libretexts.org is a site that the team uses to test out new features, it is not intended for authors. Thank you for your patience!

**Comment by pbmacsodak** [comment]
Thank you for the clarification...I am excited about the work uou all are
doing...thanks so much for the assist in getting OER into the hands of
students

**Comment by celine168** [comment]
@pbmacsodak, there should be an option to use C++ on LibreTexts now!

**Comment by moorepants** [comment]
I just tested on chem.libretexts.org and it works! Thanks everyone!

@pbmacsodak Let us know how it goes when using the tool.

**Comment by pbmacsodak** [comment]
YOU ALL ROCK!!! Thank you so much, it is soooooooo greatly appreciated. I
am starting work on a C++ text next week, so the timing is impeccable.
Pat

**Pull request merged**